tests/test_mc_simulator.py

```diff
@@ -0,0 +1,420 @@
+    test_testing_queue_item_2 = {
+        "command": "add_read_bytes",
+        "read_bytes": b"second test bytes",
+    }
+    testing_queue.put(test_testing_queue_item_2)
+    confirm_queue_is_eventually_of_size(testing_queue, 2)
+
+    test_input_item = b"some more bytes"
+    put_object_into_queue_and_raise_error_if_eventually_still_empty(
+        test_input_item, input_queue
+    )
+
+    test_output_item = b"some more bytes"
+    put_object_into_queue_and_raise_error_if_eventually_still_empty(
+        test_output_item, output_queue
+    )
+
+    test_error_item = {"test_error": "an error"}
+    put_object_into_queue_and_raise_error_if_eventually_still_empty(
+        test_error_item, error_queue
+    )
+
+    actual = simulator.hard_stop()
+    assert actual["fatal_error_reporter"] == [test_error_item]
+    assert actual["input_queue"] == [test_input_item]
+    assert actual["output_queue"] == [test_output_item]
+    assert actual["testing_queue"] == [
+        test_testing_queue_item_1,
+        test_testing_queue_item_2,
+    ]
+
+    confirm_queue_is_eventually_empty(input_queue)
+    confirm_queue_is_eventually_empty(output_queue)
+    confirm_queue_is_eventually_empty(error_queue)
+    confirm_queue_is_eventually_empty(testing_queue)
+
+
+def test_MantarrayMCSimulator__correctly_stores_time_since_initialized__in_setup_before_loop(
+    mocker,
+    mantarray_mc_simulator_no_beacon,
+):
+    _, _, _, _, simulator = mantarray_mc_simulator_no_beacon
+
+    expected_init_time = 15796649135715
+    expected_poll_time = 15880317595302
+    mocker.patch.object(  # Tanner (2/1/21): mocking perf_counter is generally a bad idea but can't think of any other way to test this
+        mc_simulator,
+        "perf_counter_ns",
+        autospec=True,
+        side_effect=[expected_init_time, expected_poll_time],
+    )
+
+    # before setup
+    assert simulator.get_cms_since_init() == 0
+
+    invoke_process_run_and_check_errors(simulator, perform_setup_before_loop=True)
+    # after setup
+    expected_dur_since_init = (
+        expected_poll_time - expected_init_time
+    ) // NANOSECONDS_PER_CENTIMILLISECOND
+    assert simulator.get_cms_since_init() == expected_dur_since_init
+
+
+def test_MantarrayMCSimulator_read__gets_next_available_bytes(
+    mantarray_mc_simulator_no_beacon,
+):
+    _, _, _, testing_queue, simulator = mantarray_mc_simulator_no_beacon
+
+    expected_bytes = b"expected_item"
+    test_item = {"command": "add_read_bytes", "read_bytes": expected_bytes}
+    put_object_into_queue_and_raise_error_if_eventually_still_empty(
+        test_item, testing_queue
+    )
+    invoke_process_run_and_check_errors(simulator)
+    actual_item = simulator.read(size=len(expected_bytes))
+    assert actual_item == expected_bytes
+
+
+def test_MantarrayMCSimulator_read__returns_empty_bytes_if_no_bytes_to_read(
+    mantarray_mc_simulator,
+):
+    _, _, _, _, simulator = mantarray_mc_simulator
+    actual_item = simulator.read(size=10)
```

Comment:
minor note: generally, we want to test the edge case, so just over the border of what would trigger the result we expect. the simplest way to test if no bytes are there is just to ask to read 1 byte